File: Data Structures & Algorithms/kth-smallest-integer-in-bst/submission-2.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    public int kthSmallest(TreeNode root, int k) {
        //what do we need to do, we need to perform dfs 
        //kth smallest basically just means kth number if we start counting from leftmost node
        //so go from 
        if(root == null){
            return 0;
        }

        int[] arr = {k, root.val};
        dfs(root,arr);
        return arr[1];
    }

    public void dfs(TreeNode root, int[] arr){
        if(root == null){
            return;
        }
        dfs(root.left,arr);
        arr[0]--;
        if(arr[0]== 0){
            arr[1]=root.val;
            return;
        }
        dfs(root.right,arr);
    }

    
}

File: Data Structures & Algorithms/word-break/submission-2.py
class Solution:
    def wordBreak(self, s: str, wordDict: List[str]) -> bool:
        

        dp = {len(s): True}

        def dfs(i):
            if i in dp:
                return dp[i]

            for w in wordDict:
                if (len(w) + i) <= len(s) and s[i:i+len(w)] == w:
                    if dfs(i+ len(w)):
                        dp[i] = True
                        return True

            dp[i] = False

            return False

        return dfs(0)
